codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -80,20 +81,16 @@
         @Override
         public ByteBuf cumulate(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf in) {
             try {
-                if (cumulation.writerIndex() > cumulation.maxCapacity() - in.readableBytes()
-                        || cumulation.refCnt() > 1 || cumulation.isReadOnly()) {
-                    // Expand cumulation (by replace it) when either there is not more room in the buffer
-                    // or if the refCnt is greater then 1 which may happen when the user use slice().retain() or
-                    // duplicate().retain() or if its read-only.
-                    //
-                    // See:
-                    // - https://github.com/netty/netty/issues/2327
-                    // - https://github.com/netty/netty/issues/1764
-                    cumulation = expandCumulation(alloc, cumulation, in);
-                } else {
-                    cumulation.writeBytes(in);
+                final int required = in.readableBytes();
+                if (required > cumulation.maxWritableBytes() ||
+                        (required > cumulation.maxFastWritableBytes() && cumulation.refCnt() > 1)) {
```

Comment:
this PR is dependent upon https://github.com/netty/netty/pull/9877 ... the only change is to remove the `isReadonly` check here.

from @njhill
> whether it's safe to rely on that behaviour in BTMD without checking .isReadOnly() (e.g. what about some "other" read only ByteBuf impl with prior writableBytes() semantics).

I would argue if there is another implementation of ReadOnly buffer which doesn't implement the `ByteBuf` interface properly it should be considered a bug and should be fixed (instead of having to accommodate for it here).